e3sm_diags/driver/utils/regrid.py

```diff
@@ -0,0 +1,635 @@
+    Replaces `e3sm_diags.utils.general.select_region`.
+    """
+    specs = REGION_SPECS[region]
+
+    lat, lon = specs.get("lat"), specs.get("lon")  # type: ignore
+
+    if lat is not None:
+        lat_dim = xc.get_dim_keys(ds[var_key], axis="Y")
+        ds = ds.sel({f"{lat_dim}": slice(*lat)})
+
+    if lon is not None:
+        lon_dim = xc.get_dim_keys(ds[var_key], axis="X")
+        ds = ds.sel({f"{lon_dim}": slice(*lon)})
+
+    return ds
+
+
+def _subset_on_arm_coord(ds: xr.Dataset, var_key: str, arm_site: str):
+    """Subset a variable in the dataset on the specified ARM site coordinate.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        The dataset.
+    var_key : str
+        The variable to subset.
+    arm_site : str
+        The ARM site.
+
+    Notes
+    -----
+    Replaces `e3sm_diags.utils.general.select_point`.
+    """
+    # TODO: Refactor this method
+    pass
+
+
+def align_grids_to_lower_res(
+    ds_a: xr.Dataset,
+    ds_b: xr.Dataset,
+    var_key: str,
+    tool: REGRID_TOOLS,
+    method: str,
+) -> Tuple[xr.DataArray, xr.DataArray]:
+    """Align the grids of two Dataset using the lower resolution of the two.
+
+    A variable has a lower resolution if it has less latitude coordinates,
+    and vice versa. If both resolutions are the same, no regridding will happen.
+
+    Parameters
+    ----------
+    ds_a : xr.Dataset
+        The first Dataset containing ``var_key``.
+    ds_b : xr.Dataset
+        The second Dataset containing ``var_key``.
+    var_key : str
+        The key of the variable in both datasets to regrid.
+    tool : {"esmf", "xesmf", "regrid2"}
+        The regridding tool to use. Note, "esmf" is accepted for backwards
+        compatibility with e3sm_diags and is simply updated to "xesmf".
+    method : str
+        The regridding method to use. Refer to [1]_ for more information on
+        these options.
+
+        esmf/xesmf options:
+          - "bilinear"
+          - "conservative"
+          - "conservative_normed"
+          - "patch"
+          - "nearest_s2d"
+          - "nearest_d2s"
+
+        regrid2 options:
+          - "conservative"
+
+    Returns
+    -------
+    Tuple[xr.DataArray, xr.DataArray]
+        A tuple of both DataArrays regridded to the lower resolution of the two.
+
+    Notes
+    -----
+    Replaces `e3sm_diags.driver.utils.general.regrid_to_lower_res`.
+
+    References
+    ----------
+    .. [1] https://xcdat.readthedocs.io/en/stable/generated/xarray.Dataset.regridder.horizontal.html
+    """
+    # TODO: Accept "esmf" as `tool` value for now because `CoreParameter`
+    # defines `self.regrid_tool="esmf"` by default and
+    # `e3sm_diags.driver.utils.general.regrid_to_lower_res()` accepts "esmf".
+    # Once this function is deprecated, we can remove "esmf" as an option here
+    # and update `CoreParameter.regrid_tool` to "xesmf"`.
+    if tool == "esmf":
+        tool = "xesmf"
+
+    lat_a = xc.get_dim_coords(ds_a[var_key], axis="Y")
```

Comment:
I'm wondering what error will be returned if user accidentally pass in unstructured datasets. I'm having this issue https://github.com/E3SM-Project/e3sm_diags/issues/165 in mind.

Reply:
xCDAT uses CF attributes and `cf_xarray` to map to the coordinates via CF attributes.

**Each axis in a variable should only have one dimension coordinate assigned to it, which is determined by the CF metadata.** 

If a variable has multiple dim coordinates per axis, xCDAT will raise an error because there isn't a way to determine which one to return back. For example, this case will happen if a variable has two sets of lat coordinates (e.g., `lat1` and `lat2`) that conflict by sharing the same CF metadata (e.g., `axis="latitude"`, `standard_name="latitude"`).

Reply:
The metadata file of an example E3SM EAM native output looks like this:
```
netcdf v2.LR.historical_0101.eam.h0.2014-12 {
dimensions:
        ncol = 21600 ;
        time = UNLIMITED ; // (1 currently)
        nbnd = 2 ;
        chars = 8 ;
        lev = 72 ;
        ilev = 73 ;

variables:
        double lat(ncol) ;
                lat:long_name = "latitude" ;
                lat:units = "degrees_north" ;
        double lon(ncol) ;
                lon:long_name = "longitude" ;
                lon:units = "degrees_east" ;
          float ANRAIN(time, lev, ncol) ;
                ANRAIN:mdims = 1 ;
                ANRAIN:units = "m-3" ;
                ANRAIN:long_name = "Average rain number conc" ;
                ANRAIN:cell_methods = "time: mean" ;
```
I guess we can try it later to see if the error info is useful enough to tell the users an native datasets is passed in..

Reply:
Thanks for this example. xCDAT should be able to handle E3SM EAM native output fine because each of the variable's axis should only have 1 dimension coordinate with the appropriate CF attributes.